Unclear `$lib` error message

When migrating from a Kit 2 app that has `$lib` imports you'll get

During development:
```
Internal server error: Cannot find module '$lib/images/github.svg' imported from ...
```

During build:
```
error during build:
Build failed with 1 error:

Error: [vite]: Rolldown failed to resolve import "$lib/images/svelte-welcome.png" from ...
```

We can probably make it more obvious about what to do here (unless it's too much trouble) by linking or instructing the user on how to move to `#lib` instead

--- a/.changeset/friendly-removed-modules.md
+++ b/.changeset/friendly-removed-modules.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+fix: explain the removal of `$lib` and `$service-worker` when their imports fail to resolve

--- a/packages/kit/src/exports/vite/index.js
+++ b/packages/kit/src/exports/vite/index.js
@@ -108,11 +108,13 @@ const options_regex = /(export\s+const\s+(prerender|csr|ssr|trailingSlash))\s*=/
 
 const removed_modules = [
 	{
+		name: '$lib',
 		pattern: /^\$lib(?:\/.*|\?.*)?$/,
 		message:
 			"`$lib` has been removed. Use `#lib` instead: https://svelte.dev/docs/kit/$lib. To keep using `$lib`, add `alias: { '$lib': 'src/lib' }` to your SvelteKit config."
 	},
 	{
+		name: '$service-worker',
 		pattern: /^\$service-worker(?:\?.*)?$/,
 		message:
 			'`$service-worker` has been removed. Use `immutable`, `assets` and `prerendered` from `$app/manifest`, `version` from `$app/env`, and `resolve(...)` from `$app/paths` instead: https://svelte.dev/docs/kit/$service-worker'
@@ -335,8 +337,17 @@ function kit({ svelte_config }) {
 				const resolved = await this.resolve(id, importer, { ...options, skipSelf: true });
 				if (resolved) return resolved;
 
-				for (const { pattern, message } of removed_modules) {
-					if (pattern.test(id)) throw stackless(message);
+				const aliases = svelte_config.kit.alias;
+				for (const { name, pattern, message } of removed_modules) {
+					if (!pattern.test(id)) continue;
+
+					// If the user re-added an alias for this module (as the migration message
+					// suggests), a failed resolution means a genuine missing file rather than
+					// use of the removed module. Let Vite report the real "not found" error
+					// instead of the misleading migration message.
+					if (name in aliases || `${name}/*` in aliases) return;
+
+					throw stackless(message);
 				}
 			}
 		},

--- a/packages/kit/src/exports/vite/index.js
+++ b/packages/kit/src/exports/vite/index.js
@@ -106,6 +106,19 @@ const enforced_config = {
 
 const options_regex = /(export\s+const\s+(prerender|csr|ssr|trailingSlash))\s*=/s;
 
+const removed_modules = [
+	{
+		pattern: /^\$lib(?:\/.*|\?.*)?$/,
+		message:
+			"`$lib` has been removed. Use `#lib` instead: https://svelte.dev/docs/kit/$lib. To keep using `$lib`, add `alias: { '$lib': 'src/lib' }` to your SvelteKit config."
+	},
+	{
+		pattern: /^\$service-worker(?:\?.*)?$/,
+		message:
+			'`$service-worker` has been removed. Use `immutable`, `assets` and `prerendered` from `$app/manifest`, `version` from `$app/env`, and `resolve(...)` from `$app/paths` instead: https://svelte.dev/docs/kit/$service-worker'
+	}
+];
+
 /** @type {Set<string>} */
 const warned = new Set();
 
@@ -315,6 +328,17 @@ function kit({ svelte_config }) {
 		name: 'vite-plugin-sveltekit-setup',
 		api: {
 			options: svelte_config
+		},
+		resolveId: {
+			filter: { id: removed_modules.map(({ pattern }) => pattern) },
+			async handler(id, importer, options) {
+				const resolved = await this.resolve(id, importer, { ...options, skipSelf: true });
+				if (resolved) return resolved;
+
+				for (const { pattern, message } of removed_modules) {
+					if (pattern.test(id)) throw stackless(message);
+				}
+			}
 		},
 
 		/**

--- a/packages/kit/test/build-errors/apps/removed-lib-import/.gitignore
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/.gitignore
@@ -1,0 +1,8 @@
+.DS_Store
+node_modules
+/build
+/.svelte-kit
+/package
+!.env
+.env.*
+!.env.example

--- a/packages/kit/test/build-errors/apps/removed-lib-import/package.json
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/package.json
@@ -1,0 +1,22 @@
+{
+	"name": "removed-lib-import",
+	"version": "0.0.1",
+	"private": true,
+	"scripts": {
+		"dev": "vite dev",
+		"build": "vite build",
+		"preview": "vite preview",
+		"prepare": "svelte-kit sync",
+		"check": "svelte-kit sync && tsc && svelte-check",
+		"check:watch": "svelte-check --watch"
+	},
+	"devDependencies": {
+		"@sveltejs/kit": "workspace:^",
+		"@sveltejs/vite-plugin-svelte": "catalog:",
+		"svelte": "catalog:",
+		"svelte-check": "catalog:",
+		"typescript": "catalog:",
+		"vite": "catalog:"
+	},
+	"type": "module"
+}

--- a/packages/kit/test/build-errors/apps/removed-lib-import/src/app.html
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/src/app.html
@@ -1,0 +1,12 @@
+<!doctype html>
+<html lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<link rel="icon" href="%sveltekit.assets%/favicon.png" />
+		<meta name="viewport" content="width=device-width, initial-scale=1" />
+		%sveltekit.head%
+	</head>
+	<body>
+		<div>%sveltekit.body%</div>
+	</body>
+</html>

--- a/packages/kit/test/build-errors/apps/removed-lib-import/src/routes/+page.svelte
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/src/routes/+page.svelte
@@ -1,0 +1,4 @@
+<script>
+	// @ts-expect-error this import intentionally fails
+	import '$lib/foo.js';
+</script>

--- a/packages/kit/test/build-errors/apps/removed-lib-import/tsconfig.json
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/tsconfig.json
@@ -1,0 +1,12 @@
+{
+	"extends": "$app/tsconfig",
+	"compilerOptions": {
+		"allowJs": true,
+		"checkJs": true,
+		"esModuleInterop": true,
+		"forceConsistentCasingInFileNames": true,
+		"resolveJsonModule": true,
+		"sourceMap": true,
+		"noEmit": true
+	}
+}

--- a/packages/kit/test/build-errors/apps/removed-lib-import/vite.config.js
+++ b/packages/kit/test/build-errors/apps/removed-lib-import/vite.config.js
@@ -1,0 +1,14 @@
+import { sveltekit } from '@sveltejs/kit/vite';
+import path from 'node:path';
+
+/** @type {import('vite').UserConfig} */
+const config = {
+	plugins: [sveltekit()],
+	server: {
+		fs: {
+			allow: [path.resolve('../../../../src')]
+		}
+	}
+};
+
+export default config;

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/.gitignore
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/.gitignore
@@ -1,0 +1,8 @@
+.DS_Store
+node_modules
+/build
+/.svelte-kit
+/package
+!.env
+.env.*
+!.env.example

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/package.json
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/package.json
@@ -1,0 +1,22 @@
+{
+	"name": "removed-service-worker-import",
+	"version": "0.0.1",
+	"private": true,
+	"scripts": {
+		"dev": "vite dev",
+		"build": "vite build",
+		"preview": "vite preview",
+		"prepare": "svelte-kit sync",
+		"check": "svelte-kit sync && tsc && svelte-check",
+		"check:watch": "svelte-check --watch"
+	},
+	"devDependencies": {
+		"@sveltejs/kit": "workspace:^",
+		"@sveltejs/vite-plugin-svelte": "catalog:",
+		"svelte": "catalog:",
+		"svelte-check": "catalog:",
+		"typescript": "catalog:",
+		"vite": "catalog:"
+	},
+	"type": "module"
+}

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/src/app.html
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/src/app.html
@@ -1,0 +1,12 @@
+<!doctype html>
+<html lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<link rel="icon" href="%sveltekit.assets%/favicon.png" />
+		<meta name="viewport" content="width=device-width, initial-scale=1" />
+		%sveltekit.head%
+	</head>
+	<body>
+		<div>%sveltekit.body%</div>
+	</body>
+</html>

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/src/routes/+page.svelte
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/src/routes/+page.svelte
@@ -1,0 +1,4 @@
+<script>
+	// @ts-expect-error this import intentionally fails
+	import '$service-worker';
+</script>

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/tsconfig.json
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/tsconfig.json
@@ -1,0 +1,12 @@
+{
+	"extends": "$app/tsconfig",
+	"compilerOptions": {
+		"allowJs": true,
+		"checkJs": true,
+		"esModuleInterop": true,
+		"forceConsistentCasingInFileNames": true,
+		"resolveJsonModule": true,
+		"sourceMap": true,
+		"noEmit": true
+	}
+}

--- a/packages/kit/test/build-errors/apps/removed-service-worker-import/vite.config.js
+++ b/packages/kit/test/build-errors/apps/removed-service-worker-import/vite.config.js
@@ -1,0 +1,14 @@
+import { sveltekit } from '@sveltejs/kit/vite';
+import path from 'node:path';
+
+/** @type {import('vite').UserConfig} */
+const config = {
+	plugins: [sveltekit()],
+	server: {
+		fs: {
+			allow: [path.resolve('../../../../src')]
+		}
+	}
+};
+
+export default config;

--- a/packages/kit/test/build-errors/removed-modules.spec.js
+++ b/packages/kit/test/build-errors/removed-modules.spec.js
@@ -1,0 +1,41 @@
+import { assert, test } from 'vitest';
+import { execSync } from 'node:child_process';
+import path from 'node:path';
+
+const timeout = 60_000;
+
+test('an unresolved $lib import explains how to migrate', { timeout }, () => {
+	try {
+		execSync('pnpm build', {
+			cwd: path.join(import.meta.dirname, 'apps/removed-lib-import'),
+			stdio: 'pipe',
+			timeout
+		});
+	} catch (err) {
+		const message = /** @type {Error} */ (err).message;
+		assert.ok(
+			message.includes('Use `#lib` instead'),
+			`received unexpected exception message ${message}`
+		);
+		return;
+	}
+	throw new Error();
+});
+
+test('an unresolved $service-worker import explains how to migrate', { timeout }, () => {
+	try {
+		execSync('pnpm build', {
+			cwd: path.join(import.meta.dirname, 'apps/removed-service-worker-import'),
+			stdio: 'pipe',
+			timeout
+		});
+	} catch (err) {
+		const message = /** @type {Error} */ (err).message;
+		assert.ok(
+			message.includes('Use `immutable`, `assets` and `prerendered` from `$app/manifest`'),
+			`received unexpected exception message ${message}`
+		);
+		return;
+	}
+	throw new Error();
+});

